spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/core/PubSubTemplate.java

```diff
@@ -90,11 +94,13 @@ public PubSubTemplate(PublisherFactory publisherFactory, SubscriberFactory subsc
 
 			@Override
 			public void onFailure(Throwable throwable) {
+				LOGGER.warn("Publishing to " + topic + " topic failed.", throwable);
```

Comment:
I think this might be OK, but the alternative is to just make `PubSubMessageHandler.handleMessageInternal` add an error callback to the `ListenableFuture` returned by `PubSubTemplate`, and log the error there if `sync == false`. WDYT?

Reply:
That's one possibility, but then `PubSubTemplate` clients wouldn't see any errors logged..

Another alternative is to log the error as debug instead of warn, but then errors still aren't easily visible. I think we can live with errors being logged twice for SI?

Reply:
I think the expectation for users of the template is that they should use the returned `ListenableFuture` to handle errors.

Reply:
I'm fine having logs here in the  `PubSubTemplate`, but for the `PubSubMessageHandler` I would suggest to inject `ListenableFutureCallback` to really react to the async results.
The similar behavior we have in the `KafkaProducerMessageHandler`: https://github.com/spring-projects/spring-integration-kafka/blob/master/src/main/java/org/springframework/integration/kafka/outbound/KafkaProducerMessageHandler.java#L259
Or https://github.com/spring-projects/spring-integration-aws/blob/master/src/main/java/org/springframework/integration/aws/outbound/KinesisMessageHandler.java#L170

Reply:
Well, I commented that way already: https://github.com/spring-cloud/spring-cloud-gcp/issues/81#issuecomment-318664280

Reply:
More specifically, are you talking about an error channel? There's a lot that's going on in the Kafka message handler. :)

Reply:
Well, let's start from the simple `ListenableFutureCallback` injection!
That should be enough for now.

Reply:
@artembilan but what would that `ListenableFutureCallback` achieve? Logging the error? We already do that in `PubSubTemplate`. Unless we want to add failure channels to `PubSubMessageHandler`, I'm not sure what `ListenableFutureCallback` would accomplish.

@meltsufin how would `PubSubTemplate` users intercept async errors in that case? I think adding a callback to the returned future wouldn't do because the call might have already returned by the time the future is returned.

Reply:
The injected `ListenableFutureCallback` by end-user will exactly do anything (s)he needs. That their injection and that is fully up to them what to do in the implementation.

Right, by default we may have `LoggingListenableFutureCallback` to really log error/success.

That doesn't matter if `ListenableFuture` has been finished or not before `Callback`  injection. The last one is called appropriately.

Reply:
PTAL hope this is what you were referring to!